File: lib/journey.ts
/**
 * lib/journey.ts — the Rolplay solution journey.
 *
 * The ecosystem is a left-to-right progression, not a flat list of modules:
 * knowledge is adopted (LMS), practised (Master Coach, Simulator), validated
 * (Certifier Coach), then sustained in the field (Second Brain). Learners are
 * assigned work in that order, so the dashboard should read in that order too.
 *
 * A client may hold any SUBSET of these. The subset is never reordered and gaps
 * are never filled in — a tenant with only Simulator + Certifier Coach sees
 * exactly those two, in this sequence.
 *
 * IMPORTANT — what this view does NOT claim. Each stage reports its OWN real
 * numbers. It is not a per-learner funnel: LMS learners are LearnWorlds user IDs
 * while practice/validation learners are bridge members, and there is no
 * verified identity join between those systems. Presenting stage-to-stage
 * drop-off would imply a join we cannot substantiate, so every figure here is
 * scoped to its own stage and labelled with its own meaning.
 *
 * SOURCE NOTE: rolplay_app's `r_simulator.category` has four values — COACH,
 * SIM, SEGMENT and a second-brain one. The fourth is deliberately IGNORED:
 * Second Brain data comes exclusively from its own token-authenticated API,
 * which is the verified source. Do not wire the category variant in.
 */

import type { Module } from './types'
import type { TranslationKey } from './translations'

/** Which phase of the journey a module belongs to (the diagram's headings). */
export type JourneyPhase =
  | 'cognitive'   // Aprendizaje Cognitivo
  | 'practice'    // Aprendizaje práctico
  | 'validation'  // Validación de conocimiento
  | 'excellence'  // Excelencia en ventas

export interface JourneyStage {
  module: Module
  phase: JourneyPhase
  /** Module display name — reuses the sidebar's nav labels so naming stays in sync. */
  labelKey: TranslationKey
  phaseKey: TranslationKey
  /** Where the stage's full detail lives. */
  href: string
  /**
   * What this stage's progress bar measures. Each stage measures a DIFFERENT
   * thing, so the UI must label it rather than implying one shared scale.
   */
  progressKey: TranslationKey
}

/**
 * Canonical order. Index in this array IS the journey position — do not sort
 * stages anywhere else, derive from here.
 */
const STAGES: readonly JourneyStage[] = [
  {
    module: 'lms',
    phase: 'cognitive',
    labelKey: 'navLms',
    phaseKey: 'journeyPhaseCognitive',
    href: '/lms',
    progressKey: 'completionRate',
  },
  {
    module: 'coach',
    phase: 'practice',
    labelKey: 'navCoach',
    phaseKey: 'journeyPhasePractice',
    href: '/coach',
    progressKey: 'colPassRate',
  },
  {
    module: 'simulator',
    phase: 'practice',
    labelKey: 'navSimulator',
    phaseKey: 'journeyPhasePractice',
    href: '/simulator',
    progressKey: 'colPassRate',
  },
  {
    module: 'certification',
    phase: 'validation',
    labelKey: 'navCertification',
    phaseKey: 'journeyPhaseValidation',
    href: '/certification',
    progressKey: 'colPassRate',
  },
  {
    module: 'second-brain',
    phase: 'excellence',
    labelKey: 'navSecondBrain',
    phaseKey: 'journeyPhaseExcellence',
    href: '/second-brain',
    progressKey: 'journeyActiveMembers',
  },
] as const

export const JOURNEY_ORDER: readonly Module[] = STAGES.map(s => s.module)

/**
 * The tenant's stages, in journey order.
 *
 * Filters STAGES by what the tenant has rather than iterating the caller's list,
 * so an out-of-order or duplicated input cannot change the sequence.
 */
export function journeyStages(available: readonly Module[]): JourneyStage[] {
  const has = new Set(available)
  return STAGES.filter(s => has.has(s.module))
}

/**
 * A journey needs at least two stages to be a journey. With one service there is
 * no progression to show, so the nav entry is hidden rather than leading to a
 * page with a single lonely card.
 */
export function hasJourney(available: readonly Module[]): boolean {
  return journeyStages(available).length >= 2
}

/** Contiguous runs of the same phase, for drawing the diagram's phase headings. */
export function journeyPhaseGroups(
  stages: readonly JourneyStage[],
): { phase: JourneyPhase; phaseKey: TranslationKey; stages: JourneyStage[] }[] {
  const groups: { phase: JourneyPhase; phaseKey: TranslationKey; stages: JourneyStage[] }[] = []
  for (const s of stages) {
    const last = groups[groups.length - 1]
    if (last && last.phase === s.phase) last.stages.push(s)
    else groups.push({ phase: s.phase, phaseKey: s.phaseKey, stages: [s] })
  }
  return groups
}

File: ai-service/app/journey.py
"""The Rolplay solution journey — Python port of lib/journey.ts's canonical
ordering, kept deliberately in sync with it (same 5 modules, same order, same
phase groupings). See that file for the full rationale: knowledge is adopted
(LMS), practised (Master Coach, Simulator), validated (Certifier Coach), then
sustained in the field (Second Brain).

A dashboard's `schema.modules` only qualifies as journey-able when every
discovered value is already one of these 5 canonical names — e.g.
rolplay_app_sql's schema_discovery maps r_simulator.category via
_CATEGORY_TO_MODULE before ever reaching here (COACH/SIM/SEGMENT -> the exact
strings below). A connector whose modules are raw, unclassified strings (e.g.
pharma_kpi's activity_type values like "Coach evaluador") is NOT run through
this — that would mean guessing a low-confidence LMS/Coach/Simulator label
with no verified mapping, which is worse than not showing a journey at all.
"""
from __future__ import annotations

CANONICAL_ORDER: list[str] = ["lms", "coach", "simulator", "certification", "second-brain"]

PHASE: dict[str, str] = {
    "lms": "cognitive",
    "coach": "practice",
    "simulator": "practice",
    "certification": "validation",
    "second-brain": "excellence",
}

LABEL: dict[str, str] = {
    "lms": "LMS",
    "coach": "Master Coach",
    "simulator": "Practice Simulator",
    "certification": "Certification",
    "second-brain": "Second Brain",
}

_CANONICAL_SET = set(CANONICAL_ORDER)

# rolplay_app_sql's r_simulator.category -> canonical module. Single source of
# truth for both schema_discovery.py (which modules a tenant has) and
# preview_fetch.py (the journey widget's per-module query) — previously
# duplicated as schema_discovery.py's own private _CATEGORY_TO_MODULE; kept
# here instead so the two can never drift apart. 'SB' (Second Brain) and
# anything else unmapped are deliberately excluded — Second Brain has its own
# dedicated, token-authenticated API and must never be double-counted through
# this one (see lib/bridge-rolplay-app.ts's matching note).
CATEGORY_TO_MODULE: dict[str, str] = {"COACH": "coach", "SIM": "simulator", "SEGMENT": "certification"}

# Reverse of the above — used to scope a per-module page's queries down to
# just that module's sessions (e.g. dashboard_planning.py's per-module pages,
# fetched via preview_fetch.py's _category_clause()).
MODULE_TO_CATEGORY: dict[str, str] = {v: k for k, v in CATEGORY_TO_MODULE.items()}


def is_canonical(modules: list[str]) -> bool:
    """True only if EVERY discovered module is a verified canonical name —
    not a majority, not a guess. One unrecognized string means the source
    isn't classifying modules the same way this ontology does, and forcing
    it in would silently misrepresent that module's real identity."""
    return bool(modules) and all(m in _CANONICAL_SET for m in modules)


def ordered_stages(modules: list[str]) -> list[str]:
    """The tenant's canonical modules, in fixed journey order — mirrors
    lib/journey.ts's journeyStages(): filters CANONICAL_ORDER by what's
    present, never sorts or reorders by the input's own order."""
    have = set(modules)
    return [m for m in CANONICAL_ORDER if m in have]


def has_journey(modules: list[str]) -> bool:
    """A journey needs >=2 stages to show a progression — mirrors
    lib/journey.ts's hasJourney()."""
    return is_canonical(modules) and len(ordered_stages(modules)) >= 2
